### PR TITLE
rpcclient: make shutdown interrupt in-flight POSTs

### PR DESCRIPTION
**Pull request by wydengyre**
## Change Description
Modify the rpcclient http POST call to ensure that a shutdown immediately interrupts in-flight requests, which otherwise would have to wait until timeout.

This PR includes the change in https://github.com/btcsuite/btcd/pull/2450 as it is necessary to ensure interruption during Dial.

## Steps to Test
The test suite has an added test to ensure this works.

### Testing
- [x] Your PR passes all CI checks.
- [x] Tests covering the positive and negative (error paths) are included.
- [x] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change is not [insubstantial](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md#substantial-contributions-only). Typo fixes are not accepted to fight bot spam.
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md#code-documentation-and-commenting) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md#model-git-commit-messages). 
- [x] Any new logging statements use an appropriate subsystem and logging level.

📝 Please see our [Contribution Guidelines](https://github.com/btcsuite/btcd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

**Review by kmk142789 — APPROVED** [review]

File: rpcclient/infrastructure.go [hunk 1]
```diff
@@ -782,12 +782,17 @@ func (c *Client) handleSendPostMessage(jReq *jsonRequest) {
...
+retryloop:
 	for i := 0; i < tries; i++ {
 		var httpReq *http.Request
 
 		bodyReader := bytes.NewReader(jReq.marshalledJSON)
-		httpReq, err = http.NewRequest("POST", httpURL, bodyReader)
+		httpReq, err = http.NewRequestWithContext(ctx, "POST", httpURL, bodyReader)
 		if err != nil {
+			// We must observe the contract that shutdown returns ErrClientShutdown.
+			if errors.Is(err, context.Canceled) && errors.Is(context.Cause(ctx), ErrClientShutdown) {
```
Comment (kcalvinalvin) [inline rpcclient/infrastructure.go]: This check is unneeded since `http.NewRequestWithContext` will never return a error that's of type `context.Canceled`.

**Comment by kcalvinalvin** [comment]
Checked:

1: The added `TestHTTPPostShutdownInterruptsPendingRequest` test is correct.
2: Sanity checked the changes in `infrastructure.go`.

One nitpicky gripe is maybe the test could be commented better.

Looks good overall except for that one unneeded check.

File: rpcclient/infrastructure.go (hunk 1 above)
Comment (wydengyre) [reply rpcclient/infrastructure.go]: Removed in https://github.com/btcsuite/btcd/pull/2451/commits/f87d91434ee7912a128a4606a9401427a70ad96a

**Comment by wydengyre** [comment]
> Checked:
> 
> 1: The added `TestHTTPPostShutdownInterruptsPendingRequest` test is correct. 2: Sanity checked the changes in `infrastructure.go`.
> 
> One nitpicky gripe is maybe the test could be commented better.
> 
> Looks good overall except for that one unneeded check.

@kcalvinalvin thanks for your review.

I've addressed your comments in https://github.com/btcsuite/btcd/pull/2451/commits/f87d91434ee7912a128a4606a9401427a70ad96a

Can you re-review?

**Comment by seeforschauer** [comment]
Hey @wydengyre, nice work on this — we hit the same pain point in production.

We run a HAProxy health checker that polls BTC nodes via `GetBlockCount()`. Each probe spawns a goroutine; when a node hangs, `GetBlockCount` blocks for ~600s (retry loop in `handleSendPostMessage` — 10 attempts x 60s HTTP timeout) while new probes keep arriving. We observed **117 leaked goroutines** on staging.

We searched the issue tracker and PRs for existing context support work — found #2450 (dial timeout, merged), this PR (shutdown interrupt), and #1323 (connection reuse), but nothing addressing per-request `context.Context` for RPC methods. The gap is clear: callers currently have no way to cancel or timeout an individual `GetBlockCount`/`SendRawTransaction`/etc. call.

Your PR fixes shutdown propagation, which is great. The transport plumbing is almost there with `http.NewRequestWithContext`. Would you or maintainers be open to extending this to per-request context support? Something like:

```go
func (c *Client) GetBlockCountWithContext(ctx context.Context) (int64, error)
```

Minimal changes: `ctx` field on `jsonRequest`, `ctx.Done()` check in the retry loop, `SendCmdCtx` variant. Non-breaking — existing methods keep working. ~80-100 lines.

Happy to open an issue + PR for this if there's interest. We worked around it with a singleflight pattern at our layer, but the proper fix belongs here.

**Comment by wydengyre** [comment]
> Would you or maintainers be open to extending this to per-request context support?

@seeforschauer I had the same problem as you, and actually have a private library that fixes it in a backward-compatible way. But I'm not affiliated with this project in any official way, so sadly the decision to support such a feature is outside my influence. I would certainly vote in favor of it, though, as it's both pragmatic/necessary and more "Go-like".

**Comment by seeforschauer** [comment]
Thanks for the support, and good to know you've solved it independently — validates the approach.

I went ahead and opened #2499 to formalize the request for per-request `context.Context` support. If you'd be open to sharing your backward-compatible approach (or collaborating on a PR), happy to coordinate there.

In the meantime, pinging @kcalvinalvin — any thoughts on merging this PR as-is and tracking context support separately in #2499?

File: rpcclient/infrastructure.go
Comment (starius) [inline rpcclient/infrastructure.go:812]: If `errors.Is(err, context.Canceled) && errors.Is(context.Cause(ctx), ErrClientShutdown) && i == tries-1`, the added error remapping logic is skipped. The code returns context.Canceled in this case, but it is expected to return ErrClientShutdown instead.

Comment (starius) [inline rpcclient/infrastructure.go:865]: If `ctx` is cancelled in the middle of `io.ReadAll` running, we return `fmt.Errorf("error reading json reply: ...`, but I think we should also return ErrClientShutdown in this case, right?

File: rpcclient/infrastructure_test.go [hunk 2]
```diff
@@ -108,3 +112,85 @@ func TestParseAddressString(t *testing.T) {
...
+	select {
+	case <-waitDone:
+	case <-time.After(5 * time.Second):
+		t.Fatalf("client shutdown did not complete")
+	}
+
+	result, err := future.Receive()
+	require.Zero(t, result)
+	require.ErrorContains(t, err, ErrClientShutdown.Error())
+}
```
Comment (starius) [inline rpcclient/infrastructure_test.go:196]: I propose to add more test coverage of the edge cases:

- Cancellation while waiting in retry backoff path (`ctx.Done()` select branch).
- Cancellation on final retry attempt (`i == tries-1`).
- Cancellation during response body read after a successful `Do`.

**Comment by starius** [comment]
I'm going to copy the commits from this PR on top of https://github.com/btcsuite/btcd/pull/2500 and resolve the remaining comments.

@wydengyre If you already addressed some of them, please push, so I have the latest version of your commits.

**Comment by wydengyre** [comment]
> @wydengyre If you already addressed some of them, please push, so I have the latest version of your commits.

I have not, sorry. Please go ahead.

**Comment by starius** [comment]
This PR was integrated into https://github.com/btcsuite/btcd/pull/2500
